src/core/qgsbookmarkmanager.cpp

```diff
@@ -0,0 +1,510 @@
+{
+  if ( id.isEmpty() )
+    return false;
+
+  int pos = -1;
+  int i = 0;
+  for ( const QgsBookmark &b : qgis::as_const( mBookmarks ) )
+  {
+    if ( b.id() == id )
+    {
+      pos = i;
+      break;
+    }
+    i++;
+  }
+
+  if ( pos < 0 )
+    return false;
+
+  emit bookmarkAboutToBeRemoved( id );
+  mBookmarks.removeAt( pos );
+  emit bookmarkRemoved( id );
+  if ( mProject )
+  {
+    mProject->setDirty( true );
+  }
+
+  return true;
+}
+
+bool QgsBookmarkManager::updateBookmark( const QgsBookmark &bookmark )
+{
+  // check for duplicate ID
+  int i = 0;
+  for ( const QgsBookmark &b : qgis::as_const( mBookmarks ) )
+  {
+    if ( b.id() == bookmark.id() )
+    {
+      mBookmarks[i] = bookmark;
+      emit bookmarkChanged( bookmark.id() );
+      if ( mProject )
+      {
+        mProject->setDirty( true );
+      }
+
+      return true;
+    }
+    i++;
+  }
+  return false;
+}
+
+void QgsBookmarkManager::clear()
+{
+  const QList< QgsBookmark > bookmarks = mBookmarks;
+  for ( const QgsBookmark &b : bookmarks )
+  {
+    removeBookmark( b.id() );
+  }
+}
+
+QList<QgsBookmark> QgsBookmarkManager::bookmarks() const
+{
+  return mBookmarks;
+}
+
+QgsBookmark QgsBookmarkManager::bookmarkById( const QString &id ) const
+{
+  for ( const QgsBookmark &b : mBookmarks )
+  {
+    if ( b.id() == id )
+      return b;
+  }
+  return QgsBookmark();
+}
+
+bool QgsBookmarkManager::readXml( const QDomElement &element, const QDomDocument &doc )
+{
+  clear();
+
+  QDomElement bookmarksElem = element;
+  if ( element.tagName() != QStringLiteral( "Bookmarks" ) )
+  {
+    bookmarksElem = element.firstChildElement( QStringLiteral( "Bookmarks" ) );
+  }
+  bool result = true;
+  if ( mProject && bookmarksElem.isNull() )
+  {
+    // handle legacy projects
+    const int count = mProject->readNumEntry( QStringLiteral( "Bookmarks" ), QStringLiteral( "/count" ) );
+    for ( int i = 0; i < count; ++i )
+    {
+      const double minX = mProject->readDoubleEntry( QStringLiteral( "Bookmarks" ), QStringLiteral( "/Row-%1/MinX" ).arg( i ) );
+      const double minY = mProject->readDoubleEntry( QStringLiteral( "Bookmarks" ), QStringLiteral( "/Row-%1/MinY" ).arg( i ) );
+      const double maxX = mProject->readDoubleEntry( QStringLiteral( "Bookmarks" ), QStringLiteral( "/Row-%1/MaxX" ).arg( i ) );
+      const double maxY = mProject->readDoubleEntry( QStringLiteral( "Bookmarks" ), QStringLiteral( "/Row-%1/MaxY" ).arg( i ) );
+      const long srid = mProject->readNumEntry( QStringLiteral( "Bookmarks" ), QStringLiteral( "/Row-%1/SRID" ).arg( i ) );
+      QgsBookmark b;
+      b.setId( QStringLiteral( "bookmark_%1" ).arg( i ) );
+      b.setName( mProject->readEntry( QStringLiteral( "Bookmarks" ), QStringLiteral( "/Row-%1/Name" ).arg( i ) ) );
+      b.setExtent( QgsReferencedRectangle( QgsRectangle( minX, minY, maxX, maxY ), QgsCoordinateReferenceSystem::fromSrsId( srid ) ) );
+
+      bool added = false;
+      addBookmark( b, &added );
```

Comment:
Is that going to trigger a bunch of signals when reading an XML containing say 50-100 bookmarks? Is this something we might not want to have happening? 


I.e., if I'm listening for project bookmark manager signals of added bookmarks, loading a new project with a bunch of bookmarks could be problematic. We might be better adding a different signal here, like bookmarksRead.

Reply:
We could add a second signal, but we need the modular signal for the model to correctly update.

Reply:
@nyalldawson , hmm, let me think here. I guess one way around being flooded with signals would be (for e.g. the browser panel) to disconnect the bookmarkAdded signal while a project is loading?